File: internal/store/taste_filter.go
package store

import "database/sql"

// tasteFilterKey is the fixed primary key of the singleton pre-filter row.
const tasteFilterKey = "default"

// GetTasteFilter returns the saved pre-filter rules (raw TOML), or "" when none
// has been saved (the caller falls back to the compiled-in default).
func (db *DB) GetTasteFilter() (string, error) {
	var c string
	err := db.QueryRow("SELECT content FROM taste_filter WHERE key = ?", tasteFilterKey).Scan(&c)
	if err == sql.ErrNoRows {
		return "", nil
	}
	return c, err
}

// PutTasteFilter upserts the singleton pre-filter row.
func (db *DB) PutTasteFilter(content string) error {
	const q = `
INSERT INTO taste_filter (key, content, updated_at)
VALUES (?, ?, CURRENT_TIMESTAMP)
ON CONFLICT(key) DO UPDATE SET content = excluded.content, updated_at = CURRENT_TIMESTAMP`
	_, err := db.Exec(q, tasteFilterKey, content)
	return err
}

File: scout/store/taste_filter.py
"""Singleton pre-filter rules row. Port of internal/store/taste_filter.go."""
from __future__ import annotations

import sqlite3

_TASTE_FILTER_KEY = "default"


def get_taste_filter(con: sqlite3.Connection) -> tuple[str, bool]:
    """Return (content, enabled) for the saved pre-filter rules. No row →
    ("", True): no rules yet (caller falls back to the compiled-in default) and
    on by default."""
    row = con.execute(
        "SELECT content, enabled FROM taste_filter WHERE key = ?", (_TASTE_FILTER_KEY,)
    ).fetchone()
    if row is None:
        return "", True
    return row[0], bool(row[1])


def put_taste_filter(con: sqlite3.Connection, content: str, enabled: bool) -> None:
    """Upsert the singleton row's rules and enabled flag together."""
    con.execute(
        """
INSERT INTO taste_filter (key, content, enabled, updated_at)
VALUES (?, ?, ?, CURRENT_TIMESTAMP)
ON CONFLICT(key) DO UPDATE SET content = excluded.content, enabled = excluded.enabled, updated_at = CURRENT_TIMESTAMP""",
        (_TASTE_FILTER_KEY, content, 1 if enabled else 0),
    )
